text/0000-let-expression.md

```diff
@@ -0,0 +1,917 @@
+
+People find let expressions theoretical at first look, and think only traditional cases (if-let-chain and let-else)
+can become useful. In this section there are some usages for new features of this RFC in real codes.
+
+This is an example from rust-clippy repository:
+```rust
+for w in block.stmts.windows(2) {
+    if_chain! {
+        if let StmtKind::Semi(first) = w[0].kind;
+        if let StmtKind::Semi(second) = w[1].kind;
+        if !differing_macro_contexts(first.span, second.span);
+        if let ExprKind::Assign(lhs0, rhs0, _) = first.kind;
+        if let ExprKind::Assign(lhs1, rhs1, _) = second.kind;
+        if eq_expr_value(cx, lhs0, rhs1);
+        if eq_expr_value(cx, lhs1, rhs0);
+        then {
+            // 30 lines of code with massive rightward drift
+        }
+    }
+}
+```
+
+Which by a generalized let-else can become:
+```rust
+for w in block.stmts.windows(2) {
+    let StmtKind::Semi(first) = w[0].kind
+    && let StmtKind::Semi(second) = w[1].kind
+    && !differing_macro_contexts(first.span, second.span)
+    && let ExprKind::Assign(lhs0, rhs0, _) = first.kind
+    && let ExprKind::Assign(lhs1, rhs1, _) = second.kind
+    && eq_expr_value(cx, lhs0, rhs1)
+    && eq_expr_value(cx, lhs1, rhs0)
+    || continue;
+    // 30 lines of code with two less tab
+}
+```
+Every `if let` or `if_chain!` that fill body of a loop or function can refactored in this way. You can easily find
+dozens of them just in rust-clippy. Note that if-let-chain alone can't solve this problem, because you have access to
+variables of if-let inside of block, but you have access to variables of a binding statement under it, thus you can
+save one indentation.
+
+`||` is not only useful for let-else style things. This is a real example from [deno](https://github.com/denoland/deno):
+
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<ChildStderrResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TcpStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<TlsStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<UnixStreamResource>() {
+    s.read(buf).await?
+} else if let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
+```
+Which by `||` counterpart of if-let-chain can become:
+```rust
+let nread = if let Some(s) = resource.downcast_rc::<ChildStdoutResource>()
+    || let Some(s) = resource.downcast_rc::<ChildStderrResource>()
+    || let Some(s) = resource.downcast_rc::<TcpStreamResource>()
+    || let Some(s) = resource.downcast_rc::<TlsStreamResource>()
+    || let Some(s) = resource.downcast_rc::<UnixStreamResource>()
+    || let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.read(buf).await?
+} else {
+    return Err(not_supported());
+};
```

Comment:
That is not possible, because those are different types. It could be done with trait objects but would be worse than the original code.

Reply:
Ah, good catch!